### DISCUSSION TITLE
tutorials/voyage-part2/08-publishing-our-schemas

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#62] [course feedback]
> **Note: The course/lesson has been updated; this discussion is outdated.**

# tutorials/voyage-part2/08-publishing-our-schemas

https://www.apollographql.com/tutorials/voyage-part2/08-publishing-our-schemas

**Comment by shareef84** [comment]
I don't get this error - Cannot return null for non-nullable field Host.name. I see a different error -  "message": "Cannot query field \"name\" on type \"User\"."

am I missing something?

**Reply by mabuyo** [reply]
Hey @shareef84! Hmmm, interesting. In Explorer, when you take a look at the fields on the left hand side under the "Documentation" panel, click into `me`, what do you see? Is there a `name` field listed there?

<img width="344" alt="Screenshot 2023-08-23 at 11 15 12 AM" src="https://github.com/apollographql-education/content-feedback/assets/6900407/83d87ecf-1c6d-4c2e-8112-34f5efaf8d3d">

**Comment by davidbuckell** [comment]
Following the failed publish of the accounts subgraph and re-publish of the monolith, are you not required to also re-publish the accounts subgraph after the successful monolith republish?
Upon reaching the 'Testing our schema changes' section I do not see the accounts subgraph in Apollo Studio SDL

**Reply by davidbuckell** [reply]
After I republish the accounts subgraph all is well, so I believe this is just a missed step in the documentation?

**Reply by lizhennessy** [reply]
Thanks for bringing this to our attention, @davidbuckell ! I'll take a look and make sure that all the instructions are updated accordingly.

**Comment by Aimeana100** [comment]
Any help for this errors when i try to start servers and publishing the monolith after creating the accounts subgraph ?? 

$ rover subgraph publish Airlock-7ncz7b@current --schema ./subgraph-accounts//schema.graphql --name accounts --routing-url http://localhost:4002

WARN: The following build errors occurred:
INVALID_GRAPHQL: [monolithic] Interface field User.name expected but Host does not provide it.
INVALID_GRAPHQL: [monolithic] Interface field User.profilePicture expected but Host does not provide it.
INVALID_GRAPHQL: [monolithic] Interface field User.name expected but Guest does not provide it.
INVALID_GRAPHQL: [monolithic] Interface field User.profilePicture expected but Guest does not provide it.

**Reply by mabuyo** [reply]
Hey @Aimeana100 - it seems that you might have missed the step where in the `monolith` subgraph, we remove fields from the `User` interface.

```diff
"Represents an Airlock user's common properties"
interface User {
  id: ID!
- "The user's first and last name"
- name: String!
- "The user's profile photo URL"
- profilePicture: String!
}
```

Can you try that?

**Comment by Dark417** [comment]
Hello team, I was able to run query in studio, GetHostListings, with bear token, now it says '"endpoint url missing". How do I reset it?

**Reply by Dark417** [reply]
resolved... I somehow stopped npm run launch, 

restarted it and everything worked

**Reply by mabuyo** [reply]
Glad you figured it out @Dark417 !

**Comment by mshah100** [comment]
Hi there,
  I have made the changes to subgraph-accounts schema and also to monolith graph and have published them.  

When I launch accounts subgraph it starts on port4002


accounts subgraph schema:

```graphql
extend schema
  @link(url: "https://specs.apollo.dev/federation/v2.0",
        import: ["@key", "@shareable"])


type Query {
  example: String
}

"Represents an Airlock user's common properties"
interface User {
  id: ID!
  "The user's first and last name"
  name: String!
  "The user's profile photo URL"
  profilePicture: String!
}

"A host is a type of Airlock user. They own listings."
type Host implements User @key(fields: "id") {
  id: ID! 
  "The user's first and last name"
  name: String! @shareable
  "The user's profile photo URL"
  profilePicture: String! @shareable
  "The host's profile bio description, will be shown in the listing"
  profileDescription: String! @shareable
}

"A guest is a type of Airlock user. They book places to stay."
type Guest implements User @key(fields: "id") {
  id: ID! 
  "The user's first and last name"
  name: String! @shareable
  "The user's profile photo URL"
  profilePicture: String! @shareable
}
```

monolith graph changed portion is as below:

```graphql
#### User accounts

"Represents an Airlock user's common properties"
interface User {
  id: ID!
}

"A host is a type of Airlock user. They own listings."
type Host implements User @key(fields: "id") {
  id: ID! 
  "The overall calculated rating for the host"
  overallRating: Float
}

"A guest is a type of Airlock user. They book places to stay."
type Guest implements User @key(fields: "id") {
  id: ID! 
  "Amount of money in the guest's wallet"
  funds: Float!
}
```

I try to run the query, my apollo explorer (localhost:4002I am getting an error:

```
  "data": {},
  "errors": [
    {
      "message": "Cannot query field \"me\" on type \"Query\".",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "stacktrace": [
          "GraphQLError: Cannot query field \"me\" on type \"Query\".",
          "    at Object.Field (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:51:13)",
          "    at Object.enter (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/graphql/language/visitor.js:301:32)",
          "    at Object.enter (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/graphql/utilities/TypeInfo.js:391:27)",
          "    at visit (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/graphql/language/visitor.js:197:21)",
          "    at validate (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/graphql/validation/validate.js:91:24)",
          "    at processGraphQLRequest (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/@apollo/server/dist/cjs/requestPipeline.js:100:57)",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async internalExecuteOperation (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/@apollo/server/dist/cjs/ApolloServer.js:634:16)",
          "    at async runHttpQuery (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/@apollo/server/dist/cjs/runHttpQuery.js:135:29)",
          "    at async runPotentiallyBatchedHttpQuery (/Users/ms243138/Documents/GraphQL/server/odyssey-voyage-II-server/subgraph-accounts/node_modules/@apollo/server/dist/cjs/httpBatching.js:37:16)"
        ]
      }
    }
  ]
}
```

Please advise on how to get around this.  Thanks

**Reply by mabuyo** [reply]
Hey @mshah100! It sounds like you're running this query:

```graphql
query GetMyProfile {
  me {
    id
    name
    profilePicture
  }
}
```
And sending it to localhost:4002 where your `accounts` subgraph is running. If you look at the schema for your `accounts` subgraph right now, do you see a `Query` type with a `me` field? No! (It only contains `example` right now). So the error `Cannot query field \"me\" on type \"Query\"."` is expected.

In this lesson, under the "Testing our schema changes" section, we specify checking _Studio_ to see our changes.

> Let's check the results in our graph on Apollo Studio. Navigate to the Schema page and the SDL tab. We'll see a new subgraph on this list called accounts! 🎉

And we run the query in Studio (to the router), not to the subgraph server.

If you run the query in Studio, and follow the rest of the section, it should work (note that we'll still get an error that we'll address in the next lesson). Hope that makes sense!

**Comment by neoVoda** [comment]
{
  "data": null,
  "errors": [
    {
      "message": "HTTP fetch failed from 'monolith': 500: Internal Server Error",
      "extensions": {
        "code": "SUBREQUEST_HTTP_ERROR",
        "service": "monolith",
        "reason": "500: Internal Server Error",
        "http": {
          "status": 500
        }
      }
    },
    {
      "message": "Context creation failed: Request failed with status code 400",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "AxiosError: Request failed with status code 400",
          "    at settle (C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\node_modules\\axios\\dist\\node\\axios.cjs:1967:12)",
          "    at IncomingMessage.handleStreamEnd (C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\node_modules\\axios\\dist\\node\\axios.cjs:3062:11)",
          "    at IncomingMessage.emit (node:events:530:35)",
          "    at endReadableNT (node:internal/streams/readable:1696:12)",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)"
        ]
      }
    }
  ],
  "extensions": {
    "valueCompletion": [
      {
        "message": "Cannot return null for non-nullable field Query.me",
        "path": []
      }
    ]
  }
}

**Reply by neoVoda** [reply]
Similar issue as before.

**Comment by MatthewMucker** [comment]
Which studio should I be running my queries in? Localhost:4000? Localhost:4001? Localhost:4002? The one published on apollographql.com?

**Reply by mabuyo** [reply]
Hey @MatthewMucker - good question. In this lesson, we published our subgraph schemas to "production" which is what lives in studio.apollographql.com. In the section `Testing our schema changes`, it says:

> Let's check the results in our graph on Apollo Studio.

So in this case, we're running the queries in studio.apollographql.com. The endpoint for that graph is the router running on localhost:4000.

(To further clarify, 4001 is where the monolith subgraph is running, and 4002 is where the accounts subgraph is running).
